### PR TITLE
Dev flash tutorial

### PR DESCRIPTION
**Pull request by lunnea**
For tutorials:
- [x] I double checked the wiring in the schematics/images
- [x] I exported the cover image without background
- [x] I used SVG format for pure vector images
- [x] I updated the metadata.json file of the tutorial
- [x] I updated the metadata.json file of the previous tutorial to contain the "next" link
- [x] I did a spell checking before requesting a review
- [x] I used title case for titles

### CODE REVIEW

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
+# Using the Flash on Portenta H7 – MCU
+
+This tutorial shows you how to use the internal flash memory of the microcontroller of Portenta H7 and other MbedOS-based Arduino boards (Nano 33 BLEs, etc.). This is Part 1 of a series of projects on how to use the flash memories on the MbedOS-based Arduino boards.
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:3]:
```suggestion
This tutorial shows you how to use the internal flash memory of the microcontroller of Portenta H7 and other MbedOS-based Arduino boards (Nano 33 BLEs, etc.). This is Part 1 of a series of projects on how to use the flash memories on the MbedOS-based Arduino boards.
```
In this tutorial we will use the internal flash memory of the microcontroller of Portenta H7 and other MbedOS-based Arduino boards (Nano 33 BLE,Nano 33 BLE Sense, etc.).

Comment (GDJose) [reply content/tutorials/portenta-h7/por-ard-flash/content.md:3]: Do we want to keep this sentence? 
This is Part 1 of a series of projects on how to use the flash memories on the MbedOS-based Arduino boards. 

Are we sure that it is gonna be a series of tutorials? and it is planned that we will release soon-ish a new tutorial related to this? If so, I would add it, if not, I would remove it

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+## (What) You Will Learn
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:5]:
```suggestion
## (What) You Will Learn
```
## What You Will Learn

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+-   (How) to use the Flash In-Application Programming Interface
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:6]:
```suggestion
-   How to use the Flash In-Application Programming Interface 
```

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
+# Using the Flash on Portenta H7 – MCU
+
+This tutorial shows you how to use the internal flash memory of the microcontroller of Portenta H7 and other MbedOS-based Arduino boards (Nano 33 BLEs, etc.). This is Part 1 of a series of projects on how to use the flash memories on the MbedOS-based Arduino boards.
+
+## (What) You Will Learn
+-   (How) to use the Flash In-Application Programming Interface 
+-   (How) to calculate the size of the flash memory
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:7]:
```suggestion
-   How to calculate the size of the flash memory 
```

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+## Required Hardware and Software
+-   Portenta H7 board (<https://store.arduino.cc/portenta-h7>)
+-   USB C cable (either USB A to USB C or USB C to USB C)
+-   Arduino IDE 1.8.10+ or Arduino Pro IDE 0.0.4+ 
+
+
+# Configuring the Development Environment
+The H7 comes with 2 Mbytes of Flash memory with read-while-write support + 1 Mbyte of RAM. To support the use of non-erasable memory, a part of the flash memory can be programmed to behave as a block device, that can operate as a place to store information. 
+
+To access the internal flash of the MCU, you will have to use the Flash In-Application Programming Interface (FlashIAP). This API creates a block device on top of the space still available on the flash after flashing your sketch. Please**,** note that **the free space** available for the FlashIAP block device **depends on** the space occupied by **the complete sketch** on the flash and a few **block alignments calculations**. You can then use a raw access API to save and load data from the block device.
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:18]:
```suggestion
To access the internal flash of the MCU, you will have to use the Flash In-Application Programming Interface (FlashIAP). This API creates a block device on top of the space still available on the flash after flashing your sketch. Please**,** note that **the free space** available for the FlashIAP block device **depends on** the space occupied by **the complete sketch** on the flash and a few **block alignments calculations**. You can then use a raw access API to save and load data from the block device.
```
To to access the internal flash of the MCU, you will have to use the [Flash In-Application Programming Interface (FlashIAP)](https://os.mbed.com/docs/mbed-os/v6.4/apis/flash-iap.html). This API creates a block device on top of the space still available on the flash after flashing your sketch. Keep in mind  that **the free space** available for the FlashIAP block device **depends on** the space occupied by **the complete sketch** on the flash plus a few **block alignments calculations**.Once the block device has been created, you can use a raw access API to save and load data on it.

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+https://os.mbed.com/docs/mbed-os/v6.4/apis/flash-iap.html
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:20]:
```suggestion

```

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:21]:
```suggestion
```

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+Please, be aware of the **flash r/w limits** while using raw/direct access: flash memories have a limited amount of write cycles. Typical flash memories can perform about 10000 writes cycles to the same block before starting to "wear out" and begin to lose the ability to retain data. **You can actually render your board useless with improper use of this example and described APIs.**
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:22]:
```suggestion
Please, be aware of the **flash r/w limits** while using raw/direct access: flash memories have a limited amount of write cycles. Typical flash memories can perform about 10000 writes cycles to the same block before starting to "wear out" and begin to lose the ability to retain data. **You can actually render your board useless with improper use of this example and described APIs.**
```
When saving and reading data from the flash memory, we need to be aware of the **flash r/w limits** while using raw/direct access: flash memories have a limited amount of write cycles. Typical flash memories can perform about 10000 writes cycles to the same block before starting to "wear out" and begin to lose the ability to retain data. **You can actually render your board useless with improper use of this example and described APIs.**

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+// Feed the RNG for later random content generation
+  srand(micros());
+
+}
+```
+
+## 3. Define the FlashIAPLimits.h 
+
+These helper functions will help you calculating limits for the FlashIAP block device. You will find them in the header file called FlashIAPLimits.h
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:70]:
```suggestion
These helper functions will help you calculating limits for the FlashIAP block device. You will find them in the header file called FlashIAPLimits.h
```
These helper functions will help you calculating limits for the FlashIAP block device. You will find them in the header file called `FlashIAPLimits.h`

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+ blockDevice.program(buffer, 0, size);
+```
+
+*De-init* the block-device when done:
+
+```
+blockDevice.deinit(); 
+```
+
+Please, remember that **the data stored** on the flash memory will be **erased** at every **sketch upload** and will only be **retained** through successive sketch executions, e.g. after **power cycling** or **resetting** the board.
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:174]:
```suggestion
Please, remember that **the data stored** on the flash memory will be **erased** at every **sketch upload** and will only be **retained** through successive sketch executions, e.g. after **power cycling** or **resetting** the board.
```
[note]

**Note :** **The data stored** on the flash memory will be **erased** at every **sketch upload** and will only be **retained** only through successive sketch executions, e.g. after **power cycling** or **resetting** the board.

[/note]

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+
+# Conclusion
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:176]:
```suggestion

```

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+  blockDevice.erase(0, size);
+  blockDevice.program(buffer, 0, size);
+
+  // Deinitialize the device
+  blockDevice.deinit();
+}
+
+void loop(){}
+```
+
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:253]:
```suggestion
# Conclusion
This tutorial has been demonstrating the use of the MCU's internal flash memory for the Arduino boards based on MbedOS (Portenta H7, Nano 33 BLEs).
```

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+*De-init* the block-device when done:
+
+```
+blockDevice.deinit(); 
+```
+
+Please, remember that **the data stored** on the flash memory will be **erased** at every **sketch upload** and will only be **retained** through successive sketch executions, e.g. after **power cycling** or **resetting** the board.
+
+# Conclusion
+This tutorial have been demonstrating the use of the MCU's internal flash memory for the Arduino boards based on MbedOS (Portenta H7, Nano 33 BLEs).
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:177]:
```suggestion

```

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+
+https://os.mbed.com/docs/mbed-os/v6.4/apis/flash-iap.html
+
+Please, be aware of the **flash r/w limits** while using raw/direct access: flash memories have a limited amount of write cycles. Typical flash memories can perform about 10000 writes cycles to the same block before starting to "wear out" and begin to lose the ability to retain data. **You can actually render your board useless with improper use of this example and described APIs.**
+
+[note]
+
+Note: **Limit** the usage of FlashIAP block devices to **once-in-a-time** read and write **operations**, for example, for managing `setup()`-time configuration parameters.
+**Please clarify. How do I do that? Lead with action. **
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:27]: @manchoz @sebromero do you know how to do this? could you add that info here?

File: content/tutorials/portenta-h7/por-ard-flash/content.md [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+  // Deinitialize the device
+  blockDevice.deinit();
+}
+
+void loop(){}
+```
+
+**Authors:** Giampaolo Mancini
+**Reviewed by:** Lenard [20.12.2020]  
+**Last revision:** -- []
```
Comment (GDJose) [inline content/tutorials/portenta-h7/por-ard-flash/content.md:256]:
```suggestion
**Last revision:** Jose Garcia [28.01.2021]
```

**Review by GDJose — CHANGES_REQUESTED** [review]

File: content/tutorials/portenta-h7/por-ard-flash/content.md (hunk 1 above)
Comment (manchoz) [reply content/tutorials/portenta-h7/por-ard-flash/content.md]: This sketch was Part 1 in respect of https://www.arduino.cc/pro/tutorials/portenta-h7/por-ard-kvs

Comment (manchoz) [reply content/tutorials/portenta-h7/por-ard-flash/content.md]: End here [Portenta_Storage_Examples](https://create.arduino.cc/editor/manchuino/c15a5e66-4c54-430d-914b-9326125db6c6/preview) you have Part 3, ie. using the QSPI Flash with Partitions and FAT, LittleFS and TDBStore.

**Pull request merged**